examples/server/json-schema-to-grammar.cpp

```diff
@@ -0,0 +1,734 @@
+                    successive_items += repeat(list_item_operator + "?", max_items - min_items - 1);
+                } else {
+                    successive_items += list_item_operator + "*";
+                }
+                string rule;
+                if (min_items == 0) {
+                    rule =  "\"[\" space ( " + item_rule_name + " " + successive_items + " )? \"]\" space";
+                } else {
+                    rule =  "\"[\" space " + item_rule_name + " " + successive_items + " \"]\" space";
+                }
+                return _add_rule(rule_name, rule);
+            }
+        } else if ((schema_type.is_null() || schema_type == "string") && schema.contains("pattern")) {
+            return _visit_pattern(schema["pattern"], rule_name);
+        } else if ((schema_type.is_null() || schema_type == "string") && regex_match(schema_format, regex("^uuid[1-5]?$"))) {
+            return _add_rule(rule_name == "root" ? "root" : schema_format, PRIMITIVE_RULES.at("uuid"));
+        } else if ((schema_type.is_null() || schema_type == "string") && DATE_RULES.find(schema_format) != DATE_RULES.end()) {
+            for (const auto& kv : DATE_RULES) {
+                _add_rule(kv.first, kv.second);
+            }
+            return schema_format + "-string";
+        } else if (schema.empty() || schema_type == "object") {
+            for (const auto& n : OBJECT_RULE_NAMES) {
+                _add_rule(n, PRIMITIVE_RULES.at(n));
+            }
+            return _add_rule(rule_name, "object");
+        } else {
+            if (!schema_type.is_string() || PRIMITIVE_RULES.find(schema_type.get<string>()) == PRIMITIVE_RULES.end()) {
+              _errors.push_back("Unrecognized schema: " + schema.dump());
+              return "";
+            }
+            // TODO: support minimum, maximum, exclusiveMinimum, exclusiveMaximum at least for zero
+            return _add_rule(rule_name == "root" ? "root" : schema_type.get<string>(), PRIMITIVE_RULES.at(schema_type.get<string>()));
+        }
+    }
+
+    void check_errors() {
+      if (!_errors.empty()) {
+        throw std::runtime_error("JSON schema conversion failed:\n" + join(_errors.begin(), _errors.end(), "\n"));
+      }
+      if (!_warnings.empty()) {
+        std::cerr << "WARNING: JSON schema conversion was incomplete: " + join(_warnings.begin(), _warnings.end(), "; ") << std::endl;
+      }
+    }
+
+    string format_grammar() {
+        stringstream ss;
+        for (const auto& kv : _rules) {
+            ss << kv.first << " ::= " << kv.second << endl;
+        }
+        return ss.str();
+    }
+};
+
+string json_schema_to_grammar(const json& schema) {
+  SchemaConverter converter([](const string&) { return json::object(); }, /* dotall= */ false);
+  auto copy = schema;
+  converter.resolve_refs(copy, "input");
+  converter.visit(copy, "");
+  converter.check_errors();
+  return converter.format_grammar();
+}
```

Comment:
Use 4-spaces indentation in C/C++ and Python sources

Reply:
Done